pyro/distributions/spanning_tree.py

```diff
@@ -0,0 +1,418 @@
+        truncated = laplacian[:-1, :-1]
+        log_det = torch.cholesky(truncated).diag().log().sum() * 2
+        return log_det + shift * (V - 1)
+
+    def log_prob(self, edges):
+        if self._validate_args:
+            if edges.dim() < 2 or edges.shape[-2:] != self.event_shape:
+                raise ValueError("Invalid edges shape: {}".format(edges.shape))
+        v1 = edges[..., 0]
+        v2 = edges[..., 1]
+        if self._validate_args:
+            assert (v1 < v2).all()
+        k = v1 + v2 * (v2 - 1) // 2
+        return self.edge_logits[k].sum(-1) - self.log_partition_function
+
+    def sample(self, sample_shape=torch.Size()):
+        if sample_shape:
+            raise NotImplementedError("SpanningTree does not support batching")
+        edges = sample_tree(self.edge_logits, **self.sampler_options)
+        assert edges.dim() >= 2 and edges.shape[-2:] == self.event_shape
+        return edges
+
+    def enumerate_support(self, expand=True):
+        trees = enumerate_spanning_trees(self.num_vertices)
+        return torch.tensor(trees, dtype=torch.long)
+
+
+################################################################################
+# Sampler implementation.
+################################################################################
+
+_cpp_module = None
+
+
+def _get_cpp_module():
+    """
+    JIT compiles the cpp_spanning_tree module.
+    """
+    global _cpp_module
+    if _cpp_module is None:
+        import os
+        from torch.utils.cpp_extension import load
+        path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "spanning_tree.cpp")
+        _cpp_module = load(name="cpp_spanning_tree",
+                           sources=[path],
+                           extra_cflags=['-O2'],
+                           verbose=True)
+    return _cpp_module
+
+
+def make_complete_graph(num_vertices, backend="python"):
+    """
+    Constructs a complete graph.
+
+    The pairing function is: ``k = v1 + v2 * (v2 - 1) // 2``
+
+    :param int num_vertices: Number of vertices.
+    :returns: a 2 x K grid of (vertex, vertex) pairs.
+    """
+    if backend == "python":
+        return _make_complete_graph(num_vertices)
+    elif backend == "cpp":
+        return _get_cpp_module().make_complete_graph(num_vertices)
+    else:
+        raise ValueError("unknown backend: {}".format(repr(backend)))
+
+
+def _make_complete_graph(num_vertices):
+    if num_vertices < 2:
+        raise ValueError('PyTorch cannot handle zero-sized multidimensional tensors')
+    V = num_vertices
+    K = V * (V - 1) // 2
+    v1 = torch.arange(V)
+    v2 = torch.arange(V).unsqueeze(-1)
+    v1, v2 = torch.broadcast_tensors(v1, v2)
+    v1 = v1.contiguous().view(-1)
+    v2 = v2.contiguous().view(-1)
+    mask = (v1 < v2)
+    grid = torch.stack((v1[mask], v2[mask]))
+    assert grid.shape == (2, K)
+    return grid
+
+
+def _remove_edge(grid, e2k, neighbors, components, e):
```

Comment:
could you please add comments defining some of the data structures you use? especially `e2k`? maybe also `neighbors` and a few others?